core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -296,17 +285,34 @@ public void registerSuggester(String key, Suggester<?> suggester) {
     }
 
     /**
-     * Register a new ScoreFunctionParser.
+     * Register a new ScoreFunctionBuilder. Registration does two things:
+     * <ul>
+     * <li>Register the {@link ScoreFunctionParser} which parses XContent into a {@link ScoreFunctionBuilder} using its {@link ParseField}
+     * </li>
+     * <li>Register the {@link Writeable.Reader} which reads a stream representation of the builder under a single name. This name must be
+     * the same name as the builder returns for getWriteableName.</li>
+     * </ul>
      */
-    public void registerFunctionScoreParser(ScoreFunctionParser<? extends ScoreFunctionBuilder> parser) {
-        for (String name: parser.getNames()) {
-            Object oldValue = functionScoreParsers.putIfAbsent(name, parser);
+    public <T extends ScoreFunctionBuilder<T>> void registerScoreFunction(ParseField parserName, ScoreFunctionParser<T> parser,
+            String writeableName, Writeable.Reader<T> reader) {
+        for (String name: parserName.getAllNamesIncludedDeprecated()) {
+            Object oldValue = scoreFunctionParsers.putIfAbsent(name, new Tuple<>(parserName, parser));
             if (oldValue != null) {
                 throw new IllegalArgumentException("Function score parser [" + oldValue + "] already registered for name [" + name + "]");
             }
         }
-        @SuppressWarnings("unchecked") NamedWriteable<? extends ScoreFunctionBuilder> sfb = parser.getBuilderPrototype();
-        namedWriteableRegistry.registerPrototype(ScoreFunctionBuilder.class, sfb);
+        namedWriteableRegistry.register(ScoreFunctionBuilder.class, writeableName, reader);
+    }
+
+    /**
+     * Lookup a {@link ScoreFunctionParser} by its name as used in XContent.
+     */
+    private ScoreFunctionParser<?> lookupScoreFunctionParser(String name, QueryParseContext context) {
+        Tuple<ParseField, ScoreFunctionParser<?>> parser = scoreFunctionParsers.get(name);
+        if (parser != null && context.parseFieldMatcher().match(name, parser.v1())) {
+            return parser.v2();
+        }
```

Comment:
deprecated names match too. match should always return true, or rather throw an exception in strict mode if you use a deprecated name. I think it should be an assert. We had the same discussion with Colin in IndicesQueriesRegistry I think :)